File: flutter_client/android/app/src/main/kotlin/dev/sparkison/tv/mpv/MpvPlayerPlugin.kt
package dev.sparkison.tv.mpv

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Method/event channel plugin for `m3u_tv/android_mpv` + `m3u_tv/android_mpv/events`.
 *
 * Modeled on `macos/Runner/MpvPlayer/MpvPlayerPlugin.swift`, with one added
 * constraint the Apple plugins don't have: the underlying
 * `dev.jdtech.mpv.MpvPlayer` is a process-wide singleton (see
 * `MpvPlayerCore`'s header comment), so at most one [MpvPlayerCore] is ever
 * active. `attachCore` fully disposes any previous core -- awaiting its
 * native mpv handle actually closing -- before creating and initializing
 * the next one, serialized through [mutex] so concurrent attach/dispose
 * calls (e.g. rapid channel switching recreating the platform view) can't
 * interleave and violate that constraint.
 */
class MpvPlayerPlugin(val context: Context, flutterEngine: FlutterEngine) :
    MethodChannel.MethodCallHandler, EventChannel.StreamHandler, MpvPlayerCore.MpvPlayerCoreDelegate {

    private val methodChannel = MethodChannel(flutterEngine.dartExecutor.binaryMessenger, METHOD_CHANNEL)
    private val eventChannel = EventChannel(flutterEngine.dartExecutor.binaryMessenger, EVENT_CHANNEL)
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private val mutex = Mutex()
    private val mainHandler = Handler(Looper.getMainLooper())

    private var activeCore: MpvPlayerCore? = null
    private var activeViewId: Int? = null
    private var eventSink: EventChannel.EventSink? = null

    init {
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
    }

    /**
     * Creates the [MpvPlayerCore] for [viewId] and returns it immediately
     * (its `surfaceView` is needed synchronously by
     * [MpvPlayerPlatformView]), while disposing any previous core and
     * initializing this one's native mpv handle asynchronously in the
     * background. `load`/control method calls that arrive before
     * initialization finishes are held by [waitForCore]/the core's own
     * internal mutex until it's ready.
     */
    fun attachCore(viewId: Int, viewContext: Context): MpvPlayerCore {
        val core = MpvPlayerCore(viewId, viewContext, this)
        scope.launch {
            mutex.withLock {
                val stale = activeCore
                if (stale != null) {
                    disposeCoreAwait(stale)
                }
                activeCore = core
                activeViewId = viewId
                initializeCoreAwait(core)
            }
        }
        return core
    }

    private suspend fun disposeCoreAwait(core: MpvPlayerCore) = suspendCancellableCoroutine<Unit> { continuation ->
        core.dispose { continuation.resume(Unit, onCancellation = null) }
    }

    private suspend fun initializeCoreAwait(core: MpvPlayerCore) = suspendCancellableCoroutine<Boolean> { continuation ->
        core.initialize { ok -> continuation.resume(ok, onCancellation = null) }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *>
        val viewId = (args?.get("viewId") as? Number)?.toInt()
        if (viewId == null) {
            result.error("bad-arguments", "viewId is required", null)
            return
        }

        when (call.method) {
            "load" -> {
                waitForCore(viewId, attemptsRemaining = 100) { core ->
                    if (core == null) {
                        result.success(
                            mapOf(
                                "ok" to false,
                                "error" to "mpv core not attached",
                                "code" to "backend_unavailable",
                            ),
                        )
                        return@waitForCore
                    }
                    @Suppress("UNCHECKED_CAST")
                    val headers = (args["headers"] as? Map<Any?, Any?>)
                        ?.entries
                        ?.mapNotNull { entry ->
                            val key = entry.key as? String
                            val value = entry.value as? String
                            if (key != null && value != null) key to value else null
                        }
                        ?.toMap()
                    core.load(
                        uri = args["uri"] as? String ?: "",
                        startPositionMs = (args["startPositionMs"] as? Number)?.toInt() ?: 0,
                        userAgent = args["userAgent"] as? String,
                        headers = headers,
                        externalSubtitles = parseExternalSubtitles(args["externalSubtitles"]),
                    )
                    result.success(mapOf("ok" to true))
                }
            }
            "play" -> {
                core(viewId)?.play()
                result.success(null)
            }
            "pause" -> {
                core(viewId)?.pause()
                result.success(null)
            }
            "seek" -> {
                core(viewId)?.seek((args["positionMs"] as? Number)?.toInt() ?: 0)
                result.success(null)
            }
            "stop" -> {
                core(viewId)?.stop()
                result.success(null)
            }
            "setAudioTrack" -> {
                core(viewId)?.setAudioTrack(args["trackId"] as? String)
                result.success(null)
            }
            "setSubtitleTrack" -> {
                core(viewId)?.setSubtitleTrack(args["trackId"] as? String)
                result.success(null)
            }
            "setPlaybackSpeed" -> {
                core(viewId)?.setPlaybackSpeed((args["speed"] as? Number)?.toDouble() ?: 1.0)
                result.success(null)
            }
            "dispose" -> {
                val existing = core(viewId)
                if (existing == null) {
                    result.success(null)
                    return
                }
                scope.launch {
                    mutex.withLock {
                        disposeCoreAwait(existing)
                        if (activeViewId == viewId) {
                            activeCore = null
                            activeViewId = null
                        }
                        result.success(null)
                    }
                }
            }
            else -> result.notImplemented()
        }
    }

    // MARK: - MpvPlayerCoreDelegate

    override fun mpvPlayerCore(core: MpvPlayerCore, event: Map<String, Any?>) {
        eventSink?.success(event)
    }

    // MARK: - EventChannel.StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    fun dispose() {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        val stale = activeCore
        activeCore = null
        activeViewId = null
        stale?.dispose {}
    }

    // MARK: - Helpers

    private fun core(viewId: Int): MpvPlayerCore? = if (activeViewId == viewId) activeCore else null

    private fun waitForCore(viewId: Int, attemptsRemaining: Int, completion: (MpvPlayerCore?) -> Unit) {
        val existing = core(viewId)
        if (existing != null) {
            completion(existing)
            return
        }
        if (attemptsRemaining <= 0) {
            completion(null)
            return
        }
        mainHandler.postDelayed({ waitForCore(viewId, attemptsRemaining - 1, completion) }, 20)
    }

    /**
     * Parses the `externalSubtitles` list `PlaybackSource` sends over the
     * `load` method call (see `MpvNativeBackendBase.load` in
     * lib/playback/mpv_native_backend_base.dart) into the triples
     * `MpvPlayerCore.load(externalSubtitles:)` expects: (uri, title, language).
     */
    private fun parseExternalSubtitles(raw: Any?): List<Triple<String, String?, String?>> {
        val list = raw as? List<*> ?: return emptyList()
        return list.mapNotNull { entry ->
            val map = entry as? Map<*, *> ?: return@mapNotNull null
            val uri = map["uri"] as? String
            if (uri.isNullOrEmpty()) return@mapNotNull null
            Triple(uri, map["title"] as? String, map["language"] as? String)
        }
    }

    companion object {
        const val METHOD_CHANNEL = "m3u_tv/android_mpv"
        const val EVENT_CHANNEL = "m3u_tv/android_mpv/events"
    }
}

File: flutter_client/macos/Runner/MpvPlayer/MpvPlayerPlugin.swift
// Method/event channel plugin for `m3u_tv/mac_mpv` + `m3u_tv/mac_mpv/events`.
//
// Owns a registry of MpvPlayerCore instances keyed by the Dart-generated
// `viewId` (see lib/playback/mac_mpv_native_backend.dart). The core for a
// given viewId is created by MpvPlayerPlatformViewFactory when Flutter
// instantiates the AppKitView, which can race with the Dart adapter's
// `load()` method call arriving first -- `load` retries briefly if the core
// isn't registered yet rather than failing immediately.

import FlutterMacOS
import AppKit

final class MpvPlayerPlugin: NSObject, FlutterStreamHandler, MpvPlayerCoreDelegate {
  static let methodChannelName = "m3u_tv/mac_mpv"
  static let eventChannelName = "m3u_tv/mac_mpv/events"

  private var cores: [Int: MpvPlayerCore] = [:]
  private var eventSink: FlutterEventSink?
  private let lock = NSLock()

  func attachCore(viewId: Int, to view: NSView) {
    lock.lock()
    let staleCore = cores[viewId]
    lock.unlock()
    // A platform view being recreated under a reused viewId before the old
    // one's async dispose completed would otherwise silently orphan its
    // core -- it'd keep running (and holding its native resources) with no
    // reference left in `cores` to ever dispose it.
    if let staleCore {
      staleCore.dispose {}
    }

    let core = MpvPlayerCore(viewId: viewId)
    core.delegate = self
    lock.lock()
    cores[viewId] = core
    lock.unlock()
    core.attach(to: view)
  }

  func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any], let viewId = (args["viewId"] as? NSNumber)?.intValue else {
      result(FlutterError(code: "bad-arguments", message: "viewId is required", details: nil))
      return
    }

    switch call.method {
    case "load":
      waitForCore(viewId: viewId, attemptsRemaining: 100) { [weak self] core in
        guard let core else {
          result(["ok": false, "error": "mpv core not attached", "code": "backend_unavailable"])
          return
        }
        core.load(
          uri: args["uri"] as? String ?? "",
          title: args["title"] as? String,
          startPositionMs: (args["startPositionMs"] as? NSNumber)?.intValue ?? 0,
          isLive: args["isLive"] as? Bool ?? false,
          userAgent: args["userAgent"] as? String,
          headers: args["headers"] as? [String: String],
          externalSubtitles: Self.parseExternalSubtitles(args["externalSubtitles"])
        )
        result(["ok": true])
      }
    case "play":
      core(for: viewId)?.play()
      result(nil)
    case "pause":
      core(for: viewId)?.pause()
      result(nil)
    case "seek":
      let positionMs = (args["positionMs"] as? NSNumber)?.intValue ?? 0
      core(for: viewId)?.seek(positionMs: positionMs)
      result(nil)
    case "stop":
      core(for: viewId)?.stop()
      result(nil)
    case "setAudioTrack":
      core(for: viewId)?.setAudioTrack(trackId: args["trackId"] as? String)
      result(nil)
    case "setSubtitleTrack":
      core(for: viewId)?.setSubtitleTrack(trackId: args["trackId"] as? String)
      result(nil)
    case "setPlaybackSpeed":
      let speed = (args["speed"] as? NSNumber)?.doubleValue ?? 1.0
      core(for: viewId)?.setPlaybackSpeed(speed)
      result(nil)
    case "setVolume":
      let volume = (args["volume"] as? NSNumber)?.doubleValue ?? 100.0
      core(for: viewId)?.setVolume(volume)
      result(nil)
    case "dispose":
      guard let existingCore = core(for: viewId) else {
        result(nil)
        return
      }
      // Only drop the strong reference (`cores`) once teardown has actually
      // finished -- see MpvPlayerCore.dispose's header comment for why.
      existingCore.dispose { [weak self] in
        self?.lock.lock()
        self?.cores.removeValue(forKey: viewId)
        self?.lock.unlock()
        result(nil)
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - MpvPlayerCoreDelegate

  func mpvPlayerCore(_ core: MpvPlayerCore, didEmit event: [String: Any]) {
    DispatchQueue.main.async { [weak self] in
      self?.eventSink?(event)
    }
  }

  // MARK: - FlutterStreamHandler

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  // MARK: - Helpers

  private func core(for viewId: Int) -> MpvPlayerCore? {
    lock.lock()
    defer { lock.unlock() }
    return cores[viewId]
  }

  private func waitForCore(viewId: Int, attemptsRemaining: Int, completion: @escaping (MpvPlayerCore?) -> Void) {
    if let core = core(for: viewId) {
      completion(core)
      return
    }
    if attemptsRemaining <= 0 {
      completion(nil)
      return
    }
    DispatchQueue.main.asyncAfter(deadline: .now() + 0.02) { [weak self] in
      self?.waitForCore(viewId: viewId, attemptsRemaining: attemptsRemaining - 1, completion: completion)
    }
  }

  /// Parses the `externalSubtitles` list `PlaybackSource` sends over the
  /// `load` method call (see `MpvNativeBackendBase.load` in
  /// lib/playback/mpv_native_backend_base.dart) into the tuple shape
  /// `MpvPlayerCore.load(externalSubtitles:)` expects.
  static func parseExternalSubtitles(_ raw: Any?) -> [(uri: String, title: String?, language: String?)] {
    guard let list = raw as? [[String: Any]] else { return [] }
    return list.compactMap { entry in
      guard let uri = entry["uri"] as? String, !uri.isEmpty else { return nil }
      return (uri: uri, title: entry["title"] as? String, language: entry["language"] as? String)
    }
  }
}
